ui/src/pages/namespace/Explorer/Service/ServiceEditor/Form/index.tsx

```diff
@@ -0,0 +1,153 @@
+import {
+  Controller,
+  DeepPartialSkipArrayKey,
+  UseFormReturn,
+  useForm,
+} from "react-hook-form";
+import {
+  Select,
+  SelectContent,
+  SelectGroup,
+  SelectItem,
+  SelectLabel,
+  SelectTrigger,
+  SelectValue,
+} from "~/design/Select";
+import { ServiceFormSchema, ServiceFormSchemaType } from "../schema";
+
+import EnvForm from "./EnvForm";
+import { FC } from "react";
+import { Fieldset } from "~/components/Form/Fieldset";
+import Input from "~/design/Input";
+import { useTranslation } from "react-i18next";
+import { zodResolver } from "@hookform/resolvers/zod";
+
+type FormProps = {
+  defaultConfig?: ServiceFormSchemaType;
+  children: (args: {
+    formControls: UseFormReturn<ServiceFormSchemaType>;
+    formMarkup: JSX.Element;
+    values: DeepPartialSkipArrayKey<ServiceFormSchemaType>;
+  }) => JSX.Element;
+};
+
+export const Form: FC<FormProps> = ({ defaultConfig, children }) => {
+  const { t } = useTranslation();
+  const formControls = useForm<ServiceFormSchemaType>({
+    resolver: zodResolver(ServiceFormSchema),
+    defaultValues: {
+      ...defaultConfig,
+    },
+  });
+
+  const fieldsInOrder = ServiceFormSchema.keyof().options;
+
+  const values = fieldsInOrder.reduce(
+    (object, key) => ({ ...object, [key]: formControls.watch(key) }),
+    {}
+  );
+
+  const { register, control } = formControls;
+
+  const sizeValues = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
```

Comment:
I agree both should be defined in the Schema file and I have moved the array there in 1f12bb1cb6fc39b7dac1399cb58ffcc777276810.

Regarding a zod enum, there is a problem. For the select element to work, the array values need to be strings. But in the file, the value is treated as a number and zod will not allow enums consisting of numbers. 

We could add code to generate both the array and zod number constraints from min and max variables instead of hard-coding the array, but that feels like over-engineering the problem, or what do you think?

Reply:
Okay, I'm super fine with leaving it as it is then. I thought there might be some ts side effects like this